sru/csrc/sru_cuda_kernel.cu

```diff
@@ -22,6 +22,378 @@ __forceinline__ __device__ scalar_t calc_grad_activation(int type, scalar_t x)
+        } 
+        //else {
+        //    *hp = 0;  // output 0 for a pad token
+        //}
+        *cp = cur;  // useful for backward
+        up += ncols_u;
+        cp += ncols;
+        hp += ncols;
+        xp += ncols_x;
+        pad_p = mask_pad ? (pad_p + batch) : NULL;
+    }
+}
+
+template <typename scalar_t>
+__global__ void sru_cuda_backward_kernel_simple(
+                        scalar_t* __restrict__ grad_u,
+                        scalar_t* __restrict__ grad_x,
+                        scalar_t* __restrict__ grad_wc,
+                        scalar_t* __restrict__ grad_bias,
+                        scalar_t* __restrict__ grad_init,
+                        const scalar_t* __restrict__ u,
+                        const scalar_t* __restrict__ x,
+                        const scalar_t* __restrict__ weight_c,
+                        const scalar_t* __restrict__ bias,
+                        const scalar_t* __restrict__ init,
+                        const scalar_t* __restrict__ mask_c,
+                        const unsigned char * __restrict__ mask_pad,
+                        const scalar_t* __restrict__ c,
+                        const scalar_t* __restrict__ grad_h,
+                        const scalar_t* __restrict__ grad_last,
+                        const int len,
+                        const int batch,
+                        const int d)
+{
+    const int ncols = batch*d;
+    const int col = blockIdx.x * blockDim.x + threadIdx.x;
+    if (col >= ncols) return;
+
+    const int ncols_u = ncols * 3;
+    const int ncols_x = ncols;
+
+    const auto wc1 = *(weight_c + (col%d));
+    const auto wc2 = *(weight_c + (col%d) + d);
+
+    const auto bias1 = *(bias + (col%d));
+    const auto bias2 = *(bias + (col%d) + d);
+    const auto mask = (mask_c == NULL) ? (scalar_t)1.f : (*(mask_c + col));
+    scalar_t gwc1 = 0;
+    scalar_t gwc2 = 0;
+    scalar_t gbias1 = 0;
+    scalar_t gbias2 = 0;
+    auto cur = *(grad_last + col);
+
+    const auto* up = u + (col * 3) + (len-1)*ncols_u;
+    const auto* xp = x + col + (len-1)*ncols;
+    const auto* cp = c + col + (len-1)*ncols;
+    const auto* ghp = grad_h + col + (len-1)*ncols;
+    const unsigned char* pad_p = (mask_pad == NULL) ? NULL : (mask_pad + (col/d) + (len-1)*batch);
+    auto* gup = grad_u + (col * 3) + (len-1)*ncols_u;
+    auto* gxp = grad_x + col + (len-1)*ncols;
+
+    for (int row = len-1; row >= 0; --row)
+    {
+        if ((pad_p == NULL) || !(*pad_p)) {
+            const auto prev_c_val = row ? (*(cp-ncols)) : (*(init+col));
+            const auto c_val = *cp;
+            const auto u0 = *up;
+            const auto u1 = *(up + 1);
+            const auto u2 = *(up + 2);
+
+            const auto x_val = *xp;
+            const auto gh_val = *ghp;
+            const auto g1 = sigmoidf(u1 + wc1*prev_c_val + bias1);
+            const auto g2 = sigmoidf(u2 + wc2*prev_c_val + bias2);
+
+            // h = c*g2 + x*(1-g2) = (c-x)*g2 + x
```

Comment:
is this dead code? can we remove it?

Reply:
oh I see, this is a description of what is happening in next few lines. Got it :+1: